### PR TITLE
[wip] attempt at solving 12527

### LINKED ISSUE #12527: solveset mishandles domain
Giving an explicit domain gives no solution:
```
>>> solveset(1/tan(x),x,Interval(1,10))
EmptySet()
```

whereas giving no domain and intersecting the result gives solutions

```
>>> solveset(1/tan(x),x).intersect(Interval(1,10))
{pi/2, 3*pi/2, 5*pi/2}
```

### PR DESCRIPTION
**Pull request by smichr**
Allow the solver to have full-domain freedom to find solution and only clip solution afterwards in the calling program.

#### References to other Issues or PRs

fix #12527 attempt

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -1003,8 +1003,12 @@ def _solveset(f, symbol, domain, _check=False):
...
 
     # assign the solvers to use
-    solver = lambda f, x, domain=domain: _solveset(f, x, domain)
-    inverter = lambda f, rhs, symbol: _invert(f, rhs, symbol, domain)
+    if domain.is_subset(S.Reals):
+	    solver = lambda f, x, domain=domain: _solveset(f, x, S.Reals).intersection(domain)
+		inverter = lambda f, rhs, symbol: invert_real(f, rhs, symbol).intersection(domain)
+    else:
+	    solver = lambda f, x, domain=domain: _solveset(f, x).intersection(domain)
+		inverter = lambda f, rhs, symbol: invert_complexes(f, rhs, symbol).intersection(domain)
```
Comment (smichr) [inline sympy/solvers/solveset.py:1011]:
```suggestion
    if domain.is_subset(S.Reals):
        solver = lambda f, x, domain=domain: _solveset(f, x, S.Reals).intersection(domain)
        inverter = lambda f, rhs, symbol: invert_real(f, rhs, symbol).intersection(domain)
    else:
        solver = lambda f, x, domain=domain: _solveset(f, x).intersection(domain)
        inverter = lambda f, rhs, symbol: invert_complexes(f, rhs, symbol).intersection(domain)
```

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -1003,8 +1003,12 @@ def _solveset(f, symbol, domain, _check=False):
...
 
     # assign the solvers to use
-    solver = lambda f, x, domain=domain: _solveset(f, x, domain)
-    inverter = lambda f, rhs, symbol: _invert(f, rhs, symbol, domain)
+    if domain.is_subset(S.Reals):
+        solver = lambda f, x, domain=domain: _solveset(f, x, S.Reals).intersection(domain)
+        inverter = lambda f, rhs, symbol: invert_real(f, rhs, symbol).intersection(domain)
+    else:
+        solver = lambda f, x, domain=domain: _solveset(f, x).intersection(domain)
+        inverter = lambda f, rhs, symbol: invert_complexes(f, rhs, symbol).intersection(domain)
```
Comment (smichr) [inline sympy/solvers/solveset.py:1011]:
```suggestion
        inverter = lambda f, rhs, symbol: invert_complex(f, rhs, symbol).intersection(domain)
```

**Comment by oscargus** [comment]
I played around with this a bit earlier today and would just like to point out that the solver do find the "correct" three solutions, but that these are removed by the `check_domain`-function. But that you probably have  already noticed.

**Comment by smichr** [comment]
> But that you probably have already noticed.

no, but thanks for the info. Let's see how this does.

/c

File: sympy/solvers/tests/test_solveset.py [hunk 3]
```diff
@@ -3064,3 +3064,8 @@ def test_issue_19144():
 def test_issue_19814():
     assert nonlinsolve([ 2**m - 2**(2*n), 4*2**m - 2**(4*n)], m, n
                       ) == FiniteSet((log(2**(2*n))/log(2), S.Complexes))
+
+
+def test_issue_12527():
+    ans = FiniteSet(*[i*pi/2 for i in range(1, 3, 5)])
+    assert solveset(1/tan(x), x, Interval(1,10)) == ans
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py:3071]:
```suggestion
    assert solveset(1/tan(x), x, Interval(1,10)) == ans


def test_issue_17579():
    assert solveset(sin(log(x)), x, S.Reals).contains(1)
    assert not solveset(sin(log(x)), x, Interval.Ropen(0, 1))).contains(1)
```

File: sympy/solvers/tests/test_solveset.py [hunk 4]
```diff
@@ -3064,3 +3064,13 @@ def test_issue_19144():
...
+
+
+def test_issue_17579():
+    assert solveset(sin(log(x)), x, S.Reals).contains(1)
+    assert not solveset(sin(log(x)), x, Interval.Ropen(0, 1))).contains(1)
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py:3076]:
```suggestion
    assert not solveset(sin(log(x)), x, Interval.Ropen(0, 1)).contains(1)
```

File: sympy/solvers/tests/test_solveset.py [hunk 5]
```diff
@@ -3064,3 +3064,13 @@ def test_issue_19144():
 def test_issue_19814():
     assert nonlinsolve([ 2**m - 2**(2*n), 4*2**m - 2**(4*n)], m, n
                       ) == FiniteSet((log(2**(2*n))/log(2), S.Complexes))
+
+
+def test_issue_12527():
+    ans = FiniteSet(*[i*pi/2 for i in range(1, 3, 5)])
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py:3070]:
```suggestion
    ans = FiniteSet(*[i*pi/2 for i in (1, 3, 5)])
```

**Comment by smichr** [comment]
So I see why this won't work: some routines give limited solutions that depend on the domain -- like `_modular` which only deals with Integers. If you pass Reals you will get a result in terms of Integers but if you assumed that these were the solutions considering all reals, I suspect you would be wrong. So I guess each routine has to get the actual domain but then judiciously choose whether to pass it or the superset on for further processing.

**Pull request closed without merging**